### PR TITLE
Refactor data filter functionality

### PR DESCRIPTION
**Pull request by nitbharambe**
### Changes proposed in this PR include:

Add refactoring change from step 3 4
This PR then includes moving around functions and introduction of the Ellipsis. Serialization to be handled in https://github.com/PowerGridModel/power-grid-model/pull/687 separately.

### CODE REVIEW

File: src/power_grid_model/_utils.py [hunk 1]
```diff
@@ -299,40 +301,148 @@ def convert_single_dataset_to_python_single_dataset(data: SingleDataset) -> Sing
...
+        ValueError: when the type for data_filter is incorrect
+        KeyError: with "unknown component types" for any unknown components
+        KeyError: with "unknown attributes" for unknown attribute(s) for a known component
+    """
+    dataset_meta = power_grid_meta_data[dataset_type]
+
+    for source, components in {"data_filter": data_filter.keys(), "data": available_components}.items():
+        unknown_components = [x for x in components if x not in dataset_meta]
+        if unknown_components:
+            raise KeyError(f"You have specified some unknown component types: {unknown_components} in {source}")
```
Comment (mgovers) [inline src/power_grid_model/_utils.py:424]: better to be explicit:
```suggestion
            raise KeyError(f"The following specified component types are unknown:{unknown_components} in {source}")
```
idem below

File: src/power_grid_model/_utils.py [hunk 2]
```diff
@@ -299,40 +301,148 @@ def convert_single_dataset_to_python_single_dataset(data: SingleDataset) -> Sing
...
+    unknown_attributes = {}
+    for comp_name, attrs in data_filter.items():
+        if attrs is None or attrs is Ellipsis:
+            continue
+        attr_names = dataset_meta[comp_name].dtype.names
+        diff = set(attrs).difference(attr_names) if attr_names is not None else set(attrs)
+        if diff != set():
+            unknown_attributes[comp_name] = diff
+
+    if unknown_attributes:
```
Comment (mgovers) [inline src/power_grid_model/_utils.py:435]: idem

File: src/power_grid_model/_utils.py (hunk 1 above)
Comment (nitbharambe) [reply src/power_grid_model/_utils.py:424]: The existing error message is f"You have specified some unknown component types: {unknown_components}"
Maybe its better to conform with that

File: src/power_grid_model/_utils.py [hunk 3]
```diff
@@ -299,40 +301,148 @@ def convert_single_dataset_to_python_single_dataset(data: SingleDataset) -> Sing
...
+        raise KeyError(f"You have specified some unknown attributes: {unknown_attributes}")
```
Comment (nitbharambe) [inline src/power_grid_model/_utils.py:436]:
```suggestion
        raise KeyError(f"The following specified attributes are unknown: {unknown_attributes} in data_filter")
```

File: tests/unit/test_internal_utils.py [hunk 4]
```diff
@@ -429,6 +430,53 @@ def test_convert_batch_dataset_to_batch_list_invalid_type_sparse(_mock: MagicMoc
...
+@pytest.mark.parametrize(
+    ("data_filter", "available_components", "error", "match"),
+    [
+        ({"abc": 3, "def": None}, None, ValueError, "Invalid filter provided"),
+        ({"abc": None, "def": None}, None, KeyError, "unknown component"),
+        ({"abc": None, CT.sym_load: None}, None, KeyError, "unknown component"),
+        ({"abc": ["xyz"], CT.sym_load: None}, None, KeyError, "unknown component"),
+        ({CT.node: ["xyz"], CT.sym_load: None}, None, KeyError, "unknown attributes"),
+        ({CT.node: ["xyz1"], CT.sym_load: ["xyz2"]}, None, KeyError, "unknown attributes"),
+        ({CT.node: None, CT.sym_load: None}, [CT.node, "ghi"], KeyError, "unknown component"),
```
Comment (nitbharambe) [inline tests/unit/test_internal_utils.py:466]:
```suggestion
        ({"abc": 3, "def": None}, None, ValueError, "Invalid filter provided"),
        ({"abc": None, "def": None}, None, KeyError, "components are unknown"),
        ({"abc": None, CT.sym_load: None}, None, KeyError, "components are unknown"),
        ({"abc": ["xyz"], CT.sym_load: None}, None, KeyError, "unknown component"),
        ({CT.node: ["xyz"], CT.sym_load: None}, None, KeyError, "attributes are unknown"),
        ({CT.node: ["xyz1"], CT.sym_load: ["xyz2"]}, None, KeyError, "attributes are unknown"),
        ({CT.node: None, CT.sym_load: None}, [CT.node, "ghi"], KeyError, "components are unknown"),
```

File: tests/unit/test_internal_utils.py [hunk 5]
```diff
@@ -429,6 +430,53 @@ def test_convert_batch_dataset_to_batch_list_invalid_type_sparse(_mock: MagicMoc
...
+@pytest.mark.parametrize(
+    ("data_filter", "available_components", "error", "match"),
+    [
+        ({"abc": 3, "def": None}, None, ValueError, "Invalid filter provided"),
+        ({"abc": None, "def": None}, None, KeyError, "components are unknown"),
+        ({"abc": None, CT.sym_load: None}, None, KeyError, "components are unknown"),
+        ({"abc": ["xyz"], CT.sym_load: None}, None, KeyError, "unknown component"),
+        ({CT.node: ["xyz"], CT.sym_load: None}, None, KeyError, "attributes are unknown"),
+        ({CT.node: ["xyz1"], CT.sym_load: ["xyz2"]}, None, KeyError, "attributes are unknown"),
+        ({CT.node: None, CT.sym_load: None}, [CT.node, "ghi"], KeyError, "components are unknown"),
```
Comment (nitbharambe) [inline tests/unit/test_internal_utils.py:466]:
```suggestion
        ({"abc": 3, "def": None}, None, ValueError, "Invalid filter provided"),
        ({"abc": None, "def": None}, None, KeyError, "component types are unknown"),
        ({"abc": None, CT.sym_load: None}, None, KeyError, "component types are unknown"),
        ({"abc": ["xyz"], CT.sym_load: None}, None, KeyError, "component types are unknown"),
        ({CT.node: ["xyz"], CT.sym_load: None}, None, KeyError, "attributes are unknown"),
        ({CT.node: ["xyz1"], CT.sym_load: ["xyz2"]}, None, KeyError, "attributes are unknown"),
        ({CT.node: None, CT.sym_load: None}, [CT.node, "ghi"], KeyError, "component types are unknown"),
```

**Review by mgovers — APPROVED** [review]

**Pull request merged**